core/src/main/java/com/linecorp/armeria/internal/common/GracefulConnectionShutdownHandler.java

```diff
@@ -0,0 +1,98 @@
+package com.linecorp.armeria.internal.common;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+/**
+ * Abstract class that's used implement protocol-specific graceful connection shutdown logic.
```

Comment:
- nit:
  ```suggestion
   * Abstract class that's used to implement protocol-specific graceful connection shutdown logic.
  ```
- How about adding some sentences about thread safety, such as that this class must be always invoked from the channel's event loop thread?